Refactor to use os.UserHomeDir()

gleich opened the pull request:
Part of #2470

gleich committed:
♻️ Refactor to use os.UserHomeDir()

mislav reviewed (changes requested):
Thanks, this is a good start! 

For the sake of backwards compatibility, however, we should keep the dependency on go-homedir as part of the migration path. In the original issue, I stressed this out:
> It's important to note that whenever we add new logic about config file storage, we need to ensure a migration path for users who already have configuration files in the old location.

So, whenever we try to read from a config file, we should check both the new location and the old location, and automatically move the file from old to new location if found. This migratory logic shall stay around until the next major release in the future, when we will remove it.

See https://github.com/cli/cli/pull/2429#discussion_r525999291 for more context.

gleich commented:
Ok so I just made the following changes:
1. Get the home dir using `go-homedir` and using `os`.
2. If they are not the same then move the old path to the new path. It also makes sure that the config in that old path exists and is a folder.
3. If they are the same just return the new path.

mislav commented on internal/config/config_file.go:
```diff
@@ -30,6 +31,34 @@ func ParseDefaultConfig() (Config, error) {
 
+// Get the absolute path for a folder or file. Support for the old homedir method
+// Tied to https://github.com/cli/cli/pull/2740
+func GetAbsolutePath(elem ...string) (string, error) {
```
Could this be named `HomeDirPath`?

Also, could the name somehow reflect that auto-renaming will happen if the old home location was found? Basically, the idea is that no new callers should ever use this function. It's only to be used in places where we currently relied on `mitchellh/go-homedir`.

mislav commented on internal/config/config_file.go:
```diff
@@ -30,6 +31,34 @@ func ParseDefaultConfig() (Config, error) {
+	fullOldPath := filepath.Join(append([]string{oldHomeDir}, elem...)...)
+	fullPath := filepath.Join(append([]string{homeDir}, elem...)...)
+
+	if fullOldPath != fullPath {
```
Could we avoid renaming if `elem` was empty? This is to have a guard against ever renaming the literal user's home directory, which of course we never want to move/rename.

mislav reviewed (changes requested):
This looks good! If you have time, I've requested some changes that will make the logic safer & clearer.

gleich commented:
Ok so I changed the following:
1. Added a godoc comment to the function that tells new contributors not to use this function and to instead use `os.UserHomeDir()`.
2. Checked to make sure `elem` has a length greater than one before renaming the dir.
3. Renamed the function to `DeprecatedHomeDirPath` so it is a little more clear.

mislav committed:
Avoid ever invoking go-homedir when config was found in a new location

mislav reviewed (approved):
Thank you! This looks great.

I've pushed an optimization that avoids ever running go-homedir functionality until the fallback path is activated.
